lib/parser/README.md

```diff
@@ -30,6 +45,183 @@ Unlike `parse`, this can be used to parse the entire response object.
 
 **Returns:** `object`
 
+## ObservedArray
+You may use `ObservedArray<T extends YTNode>` as a normal array, but it provides additional methods for typesafe access and casting.
+
+```ts
+// for example, we have a feed, and want all the videos
+const feed = new ObservedArray<YTNode>([...feed.contents]);
+const videos = feed.filterType(GridVideo);
+// This is now a GridVideo[]
+
+// or we want only the first video
+const firstVideo = feed.firstOfType(GridVideo);
+
+// We may cast the whole array to a GridVideo[] and throw if we have any non-GridVideo elements
+const allVideos = feed.as(GridVideo);
+
+// There's some extra methods for ObservedArray<T extends YTNode>
+// which we use internally but not documented here (yet).
+// see the source code for more details.
+```
+
+## SuperParsedResponse
+Represents a parsed response in an unknown state. Either a `YTNode` or a `ObservedArray<YTNode>` or `null`.
+
+You will need to assert the type and unwrap the response to get the actual value.
+
+```ts
+// We can assert we have a YTNode
+const response = Parser.parse(data);
+if (response.is_item) {
+  const node = response.item();
+}
+
+// We can assert we have an ObservedArray<YTNode>
+const response = Parser.parse(data);
+if (response.is_array) {
+  const nodes = response.array();
+}
+
+// or lastly a null response
+const response = Parser.parse(data);
+const is_null = response.is_null;
+```
+
+## YTNode
+All renderers returned by Innertube are converted to this generic class and then extended for the specific renderers.
+
+This class is allows us a typesafe way to use data returned by the Innertube API.
+
+Here's how to use this class to access returned data:
+
+### Type Casting
+```ts
+// We can cast a YTNode to a child class of YTNode
+const results = node.as(TwoColumnSearchResults);
+// This will throw if the node is not a TwoColumnSearchResults
+// We thus may want to check for the type of the node before casting
+if (node.is(TwoColumnSearchResults)) {
+  // We do not need to recast the node, it is already a TwoColumnSearchResults after calling is() and using it in the branch where is() returns true
+  const results = node;
+}
+
+// Sometimes we can expect multiple types of nodes
+const results = node.asOneOf<TwoColumnSearchResults | VideoList>([TwoColumnSearchResults, VideoList]);
+
+// similarly, we can check if the node is of a certain type
+if (node.isOneOf<TwoColumnSearchResults | VideoList>([TwoColumnSearchResults, VideoList])) {
+  // Again no casting is needed, the node is already of the correct type
+  const results = node;
+}
+```
```

Comment:
Just noticed this in the documentation, but the type parameters here could be inferred by typescript, if the signature was slightly modified to use varargs instead. Just a thought, mostly syntactic sugar.
```ts
function isOneOf<T extends YTNode, K extends YTNodeConstructor<T>[]>(...types: K): K[number] {
  // ...
}
```

Reply:
Ah so maybe then wrap it in a `Return<T>`? I'd have to play around with it but I think maybe that would work?

Reply:
Would be great if you could figure out how to do this. Think we can keep it like this for this PR and fix at a later date then?

Reply:
Yeah I think that's appropriate.